src/main/java/qna/domain/Question.java

```diff
@@ -1,95 +1,120 @@
-    }
-
-    public String getTitle() {
-        return title;
-    }
-
-    public Question setTitle(String title) {
-        this.title = title;
-        return this;
-    }
-
-    public String getContents() {
-        return contents;
-    }
-
-    public Question setContents(String contents) {
-        this.contents = contents;
-        return this;
-    }
-
-    public User getWriter() {
-        return writer;
-    }
-
-    public Question writeBy(User loginUser) {
-        this.writer = loginUser;
-        return this;
+  @Column(length = 100, nullable = false)
+  private String title;
+
+  @Lob private String contents;
+
+  @ManyToOne
+  @JoinColumn(foreignKey = @ForeignKey(name = "fk_question_writer"))
+  private User writer;
+
+  @OneToMany(mappedBy = "question", cascade = CascadeType.ALL)
+  @Where(clause = "deleted = false")
+  @OrderBy("id ASC")
+  private List<Answer> answers = new ArrayList<>();
+
+  private boolean deleted = false;
+
+  public Question() {}
+
+  public Question(String title, String contents) {
+    this.title = title;
+    this.contents = contents;
+  }
+
+  public Question(long id, String title, String contents) {
+    super(id);
+    this.title = title;
+    this.contents = contents;
+  }
+
+  public String getTitle() {
+    return title;
+  }
+
+  public Question setTitle(String title) {
+    this.title = title;
+    return this;
+  }
+
+  public String getContents() {
+    return contents;
+  }
+
+  public Question setContents(String contents) {
+    this.contents = contents;
+    return this;
+  }
+
+  public User getWriter() {
+    return writer;
+  }
+
+  public Question writeBy(User loginUser) {
+    this.writer = loginUser;
+    return this;
+  }
+
+  public void addAnswer(Answer answer) {
+    answer.toQuestion(this);
+    answers.add(answer);
+  }
+
+  public boolean isOwner(User loginUser) {
+    return writer.equals(loginUser);
+  }
+
+  public Question setDeleted(boolean deleted) {
+    this.deleted = deleted;
+    return this;
+  }
+
+  public boolean isDeleted() {
+    return deleted;
+  }
+
+  public List<Answer> getAnswers() {
+    return answers;
+  }
+
+  public List<DeleteHistory> delete(User loginUser) throws CannotDeleteException {
+    if (!isOwner(loginUser)) {
+      throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
     }
 
-    public void addAnswer(Answer answer) {
-        answer.toQuestion(this);
-        answers.add(answer);
-    }
-
-    public boolean isOwner(User loginUser) {
-        return writer.equals(loginUser);
-    }
+    this.deleted = true;
 
-    public Question setDeleted(boolean deleted) {
-        this.deleted = deleted;
-        return this;
-    }
+    List<DeleteHistory> deleteHistories = new ArrayList<>();
+    deleteHistories.add(new DeleteHistory(ContentType.QUESTION, id, writer, LocalDateTime.now()));
 
-    public boolean isDeleted() {
-        return deleted;
+    for (Answer answer : answers) { ;
+      deleteHistories.add(answer.delete(loginUser));
```

Comment:
이 부분은 `Stream` 이용해서 리팩토링 시도해보시면 어떨까요?

Reply:
```suggestion
    answers.forEach(answer -> answer.delete(loginUser));
```
처음에는 Stream 으로 작성을 하려고 했으나

```suggestion
    answers.forEach(
        answer -> {
          try {
            deleteHistories.add(answer.delete(loginUser));
          } catch (CannotDeleteException e) {
            throw new RuntimeException(e);
          }
        });
```
기존 소스에서 CannotDeleteException 을 Checked Exception 으로 만들어 놓아서foreach 로 구현을 하였습니다. 

Checked Exception 보다는 Unchecked Exception 리팩토링 후 Stream 으로 구현도 고민을 했지만 기존 소스를 변경하지 않으려고 했습니다.

좋은 방법이 있으면 알려주세요.

Reply:
@pcspapa 어느정도는 오버 엔지니어링 처럼 보이기는 합니다만...
https://www.slipp.net/questions/572 읽어 보시고, 한 번 시도해 보시겠어요?